
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="english">
  <head>
    <meta charset="utf-8" />
    <title>PARAMETER FILE: &#8212; ProkSeq 20.08.2020 documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="parameter-file">
<h1>PARAMETER FILE:<a class="headerlink" href="#parameter-file" title="Permalink to this headline">Â¶</a></h1>
<p>There should be one parameter file. The entries of the file should be as follows.</p>
<blockquote>
<div><p>#       File &quot;param.yaml&quot; - definition file in YAML format. Define
#       the paths and parameters to run the external packages. All
#       the default parameters are considered. However, if users wish
#       to modify any default parameters, please specify here.
#       NOTE: Any flag/parameter which do not require any value,
#       specify &quot;TRUE&quot;/&quot;FALSE.&quot; Quotes are necessary. &quot;TRUE&quot; to
#       invoke the flag and &quot;FALSE&quot; to suppress.
####################################################################
#       Describe the Bowtie options below as:
BOWTIE:</p>
<blockquote>
<div><p>-I : 0
-X : 500
-k : 1
-p : 1</p>
</div></blockquote>
<p>#       In case the package salmon is to be run, uncomment the
#       following options. These are the default values.
#       These options will overwrite the BOWTIE parameters.
#
#SALMON:
#   SALMONINDEX:
#      -k : 29
#   SALMONQUANT:
#      -l : A
#      -p : 2
#      --validateMappings : &quot;TRUE&quot;
#
#       Define the parameters for AfterQC. Example default parameters
#       are shown as bellow. All the default parameters are taken.
#       Specify the parameter and value in case of any changes. One can
#       add the other parameters with value.
#AFTERQC:
#   -s : 35
#   -n : 5
#   -p : 35
#   --debubble : &quot;TRUE&quot;
#   -a : 5
#
#       Define the Featurecounts options as bellow:
FEATURECOUNTS:</p>
<blockquote>
<div><p>a : oldAnnotationGFF.gtf #Define the Featurecounts input GTF file.
o : FeatCount #Define the Featurecounts output file name.</p>
</div></blockquote>
<p>#
#       Specify a count file name
COUNTFILE : countFile.csv
#
#       geneBody coverage.r require a bed file. Specify the name of bed file as bellow:
geneBody_coverage:</p>
<blockquote>
<div><p>r : oldAnnotationGFF.bed</p>
</div></blockquote>
<p>#
#       Specify if batch effect removal is required. FALSE if not required, else TRUE.
BATCH_EFFECT_REMOVE : &quot;FALSE&quot;
#
#Remove Unwanted Variation from RNA-Seq Data
#To run RUVSeq, uncomment the following two lines.
RUVSeq:</p>
<blockquote>
<div><p>nc : 100 #Integer - Top 100 genes as ranked by edgeR p-values. Negative control genes to estimate the factors of unwanted variation.</p>
</div></blockquote>
<p>#
#       Parameters for pathway analysis.
#       For pathway analysis, define the organism in three alphabets as bellow.
#       ypy = Yersinia pseudotuberculosis
#       User need to change the keg abbreviation of their genome which can be
#       found in <a class="reference external" href="https://www.genome.jp/kegg/catalog/org_list.html">https://www.genome.jp/kegg/catalog/org_list.html</a>. Here ypy is
#       the Yersinia pseudotuberculosis YPIII
#       If PATHWAY analysis is not required comment out the PATHWAY section.
PATHWAY:</p>
<blockquote>
<div><p>cutoffPositive : 2.0  #logfold upper limit
cutoffNegative : -2.0 #logfold lower limit
Organism : ypy
# For Gene Ontology of the pathway analysis, define GO term and gene name file.
TERM2GENE : data/TERM2GENE.csv
TERM2NAME : data/TERM2NAME.csv</p>
</div></blockquote>
<p>#
#       Specify the root path. That means where the ProkSeq bundle is unpacked.
#       The PATH ROOT : should indicate the current working directory.
#       The location should have the following folders
#       1. depend - contains all the binaries of the external packages
#       2. scripts - contains all the modules required for ProkSeq, and pipeline-vx.x.sh
#       3. data - Contains Gene ontology files for pathway analysis.
#       4. Fastq files, and other genome/transcript files.
#       If the package is stored in the folder/path /home/user/PROKSEK
PATH ROOT : /home/user/PROKSEK
#       If the above environment (depend, scripts, data) is true, the following
#       line maye uncommented.
#PATH DEFAULT : &quot;TRUE&quot;
#       Specify the path to samtools
PATH SAMTOOLS : /home/user/PROKSEK/depend/samtools/bin
#       Specify the path to geneBody_coverage
PATH geneBody_coverage : /home/user/PROKSEK/depend/RSeQC-2.6.2/scripts/
#       Specify the path to FEATURECOUNTS
PATH FEATURECOUNTS : /home/user/PROKSEK/depend/subread-1.4.6-p5-Linux-i386/bin/
#       Specify the path to fastqc
PATH FASTQC : /home/user/PROKSEK/depend/FastQC
#       Specify the path to bowtie
PATH BOWTIE : /home/user/PROKSEK/depend/bowtie2/bowtie2-2.3.5.1-linux-x86_64
#       Specify the path to salmon if salmon is required
#PATH SALMON : /home/user/PROKSEK/depend/salmon-latest_linux_x86_64/bin
#       Specify the path to pypy required for running afterqc
PATH PYPY : /home/user/PROKSEK/depend/pypy2.7-v7.2.0-linux64/bin
#       Specify the path to readfasta
PATH READFASTA : /home/user/PROKSEK/depend
#
#       End of file &quot;param.input&quot;
#</p>
</div></blockquote>
<p>In general, the entries starting with BOWTIE instructs the program to run the tool with the additional parameter. Similarly SALMON, AFTERQC, FEATURECOUNTS, etc. Entries beginning with PATH indicates the path to the executables of the external tools.  If one uses the bundled packages in the depend folder, <cite>PATH DEFAULT : TRUE</cite> line should be uncommented. Note <cite>PATH ROOT : path_to_the_current_working_directory</cite> should be mentioned.</p>
<p>NOTE: The program will override the Bowtie options, and the package salmon will be run if both bowtie and salmon's options are provided. Therefore, if salmon is required, please comment on the Bowtie option lines, and vice versa.</p>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">ProkSeq</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../PARAMETER FILE.html">PARAMETER FILE:</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, S.nandi, Firoj mahmud.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.4.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/rst/PARAMETER FILE.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>