
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="english">
  <head>
    <meta charset="utf-8" />
    <title>DOCKER: &#8212; ProkSeq 20.08.2020 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="CONDA:" href="CONDA.html" />
    <link rel="prev" title="RUNNING PROKSEQ:" href="RUNNING PROKSEQ.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="docker">
<h1>DOCKER:<a class="headerlink" href="#docker" title="Permalink to this headline">Â¶</a></h1>
<p>We strongly recommend using docker to run the pipeline. The external dependencies and R dependencies are all bundled in the container. The container prokseq-v2.1:v1 is available in https://hub.docker.com/repository/docker/snandids/prokseq-v2.1</p>
<p><strong>Step 1:</strong> To pull the image from the Docker Hub registry:</p>
<p>&gt; docker pull snandids/prokseq-v2.1:v1</p>
<p><strong>Step 2:</strong> To Run:</p>
<p>&gt; docker run -it snandids/prokseq-v2.1:v1</p>
<p>&gt; sh-5.0 # cd prokseq</p>
<p><strong>Step 3:</strong> Activate the environment</p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">sh-5.0</span></code> # source /etc/profile.d/conda.sh</p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">sh-5.0</span></code> # conda activate py36</p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">(py36)</span> <span class="pre">sh-5.0</span></code> # &lt;YOU WILL GET THIS PROMPT&gt;</p>
<p><strong>Step 4:</strong> Run the example</p>
<p>Run the pipeline with the example files:</p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">(py36)</span> <span class="pre">sh-5.0</span></code> # python scripts/prokseq.py -s samples.bowtie.PEsample -p param.bowtie.yaml -n 4</p>
<p>The script will run with PE (paired-end) samples described in samples.bowtie.PEsample, and with the parameters defined in param.input.yaml. The program is submitted with four processors.</p>
<p><strong>Step 5:</strong> How to manage data within your Docker containers.
Once the ProkSeq pipeline is working successfully with the example files, one can go for the real data.
There are two ways to manage data within your Docker containers.&lt;br/&gt;</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>A. Method 1</strong> - Using volume mounts &lt;br/&gt;</p></li>
<li><p><strong>B. Method 2</strong> - Docker cp</p></li>
</ul>
</div></blockquote>
<p><strong>A. Method 1:&lt;br/&gt;</strong>
Volumes are the preferred mechanism for persisting data generated by and used by Docker containers.&lt;br/&gt;
Exit from the existing docker and mount the container with a volume.&lt;br/&gt;</p>
<p><strong>&gt; docker run -it -v /home/user/prokseqData:/root/prokseq/ --name PROKSEQ snandids/prokseq-v2.1:v1</strong></p>
<p>Assuming the user has all the required data files (sample[fq/fastq], GTF, BED, genome/transcript files, etc) in the folder <cite>/home/user/prokseqData</cite>, the above command will mount the directory <cite>/home/user/prokseqData</cite> in <cite>/root/prokseq/</cite> inside container.&lt;br/&gt;</p>
<p>User can run the pipeline with appropriate sample files and parameter files inside the docker.</p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">sh-5.0</span></code> #cd prokseq</p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">sh-5.0</span></code> #source /etc/profile.d/conda.sh</p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">sh-5.0</span></code> #conda activate py36</p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">(py36)</span> <span class="pre">sh-5.0</span></code> # &lt;YOU WILL GET THIS PROMPT&gt;</p>
<p><strong>Modify the parameter file with appropriate path and desired parameters.&lt;br/&gt;</strong></p>
<p><strong>Modify the sample file with the fastq file names and the genome/transcript file names.</strong></p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">(py36)</span> <span class="pre">sh-5.0</span></code> #python scripts/prokseq.py -s samples_def_file -p param_def_file -n 4</p>
<p>Once done, the user can exit the docker.</p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">(py36)</span> <span class="pre">sh-5.0</span></code> # exit</p>
<p>The output will remain in the <cite>/home/user/prokseqData</cite>.</p>
<p>To remove the container.</p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">docker</span> <span class="pre">container</span> <span class="pre">rm</span> <span class="pre">PROKSEQ</span></code></p>
<dl>
<dt><strong>B. Method 2:&lt;br/&gt;</strong></dt><dd><p>To copy the files (sample[fq/fastq], GTF, BED, fasta (Genome/transcript), etc) to and from the container:&lt;br/&gt;
Find out the containerID from another terminal. For example:&lt;br/&gt;
Run the command</p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">docker</span> <span class="pre">ps</span> <span class="pre">-a</span></code></p>
</dd>
</dl>
<p><strong>Output (somewhat similar):&lt;br/&gt;</strong></p>
<p>CONTAINER ID:8f780c0a9969</p>
<p>IMAGE:snandids/prokseq-v2.1:v1</p>
<p>COMMAND:sh</p>
<p>CREATED:5 minutes ago</p>
<p>STATUS:Up 5 minutes</p>
<p>PORTS:fervent_feynman</p>
<p>Then follow the following:&lt;br/&gt;</p>
<p><strong>Copy to the container: &lt;br/&gt;</strong></p>
<p><strong>In the example below, local test.fasta file is copied to the container `(/root/prokseq/)`.</strong></p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">docker</span> <span class="pre">cp</span> <span class="pre">test.fasta</span> <span class="pre">8f780c0a9969:/root/prokseq/test.fasta</span></code></p>
<p><strong>Copy from the container:&lt;br/&gt;</strong>
In the example below, a file from <cite>/root/prokseq/</cite> will be copied to the local working directory.&lt;br/&gt;</p>
<p><strong>Step 1: First create a file in the container:</strong></p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">(py36)</span> <span class="pre">sh-5.0</span></code> # touch TEST.txt</p>
<p>OR:</p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">sh-5.0</span></code> # touch TEST.txt</p>
<p><strong>Step 2: Copy the created new file from the container to the local working directory. From another termnal run,</strong></p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">docker</span> <span class="pre">cp</span> <span class="pre">8f780c0a9969:/root/prokseq/TEST.txt</span></code>.</p>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">ProkSeq</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="INSTALLATION AND DEPENDENCIES.html">INSTALLATION AND DEPENDENCIES:</a></li>
<li class="toctree-l1"><a class="reference internal" href="REQUIREMENTS.html">REQUIRMENTS:</a></li>
<li class="toctree-l1"><a class="reference internal" href="SAMPLES FILE.html">SAMPLE FILE:</a></li>
<li class="toctree-l1"><a class="reference internal" href="SAMPLES FILE.html#data-files">DATA FILES:</a></li>
<li class="toctree-l1"><a class="reference internal" href="PARAMETER FILE.html">PARAMETER FILE:</a></li>
<li class="toctree-l1"><a class="reference internal" href="RUNNING PROKSEQ.html">RUNNING PROKSEQ:</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">DOCKER:</a></li>
<li class="toctree-l1"><a class="reference internal" href="CONDA.html">CONDA:</a></li>
<li class="toctree-l1"><a class="reference internal" href="CONDA.html#check-test-run">CHECK TEST RUN:</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="RUNNING PROKSEQ.html" title="previous chapter">RUNNING PROKSEQ:</a></li>
      <li>Next: <a href="CONDA.html" title="next chapter">CONDA:</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, S.nandi, Firoj mahmud.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.4.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/DOCKER.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>