
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="english">
  <head>
    <meta charset="utf-8" />
    <title>OUTPUT: &#8212; ProkSeq 20.08.2020 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="CONDA:" href="CONDA.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="output">
<h1>OUTPUT:<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h1>
<p>ProkSeq gives several folder with analysis results as output</p>
<p>The structure of the OUTPUT looks like below</p>
<div class="section" id="prokseq-output">
<h2>ProkSeq OUTPUT<a class="headerlink" href="#prokseq-output" title="Permalink to this headline">¶</a></h2>
<p><strong>ProkSeq</strong>..</p>
<blockquote>
<div><ol class="arabic">
<li><dl class="simple">
<dt><strong>QC_preFilter</strong>:</dt><dd><ul class="simple">
<li><p>fastQC out put html file</p></li>
</ul>
</dd>
</dl>
</li>
<li><p><strong>QC_afterFilter</strong>: Remove bad quality read and filter good quality reads. It contains three subfolder</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>FilteredReads</strong>: good quality filtered fastq files</p></li>
<li><p><strong>QCfastQ_filtered</strong> : quality checking html file of filtered reads</p></li>
<li><p><strong>RemovedReads</strong>: Bad quality reads that is removed for analysis</p></li>
</ul>
</div></blockquote>
</li>
<li><dl class="simple">
<dt><strong>sam</strong>:</dt><dd><ul class="simple">
<li><p><strong>.sam</strong>: Sequence Alignment file generated from bowtie2/salmon</p></li>
<li><p><strong>.sam.stat</strong>: ALignment statistics of individual sample.</p></li>
<li><p><strong>Alignment.summary</strong>: Summary information/statistics of alignment to the reference genome</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>TotalCountNdExpressions</strong>:</dt><dd><ul class="simple">
<li><p><strong>countFile_CPM_TPM.csv</strong> : The file contains total count according to genomic features as well as Count per miillion(CPM), and Transcript per million (TPM)</p></li>
<li><p><strong>countFile.NucleotideAvgCount.csv</strong> : THis file contains Average nucleotide expression per gene as well as other total count, CPM and TPM.</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl>
<dt><strong>DiffExpResults</strong>:</dt><dd><ul>
<li><p><strong>DESeq2_results.txt</strong>: Differential expression results from DESeq2</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>DESeq2lfcShrink_results</strong>: Differential expression results from DESeq2 with Log2 Fold Shrinkage</p></li>
</ul>
</div></blockquote>
</li>
<li><p><strong>edgeR_results.txt</strong>: Differential expression results from edgeR</p></li>
<li><p><strong>afterNoiseq.txt</strong>: Differential expression results from NOISeq</p></li>
<li><p><strong>RUV_DESeq2_results.txt</strong>: Differential expression results with RUV normalization by using RUVSeq and DESeq2</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>GenomeBrowserFile</strong>:</dt><dd><ul class="simple">
<li><p><strong>bam</strong>: Folder contain Binary alignment file (BAM) as well as  sorted and indexed BAM. Users are advised to use sorted.bam file for raw aligned file visulazation by genomic browser IGV</p></li>
<li><p><strong>.wig</strong>: Single nucleotide visualization wiggle file for visualization or other purpose</p></li>
<li><p><strong>.bw</strong> : Single nucleotide visualization Big wiggle file that takes which is memory efficient for visualization</p></li>
<li><p><strong>normalized.bw</strong>: Single nucleotide visualization normalized Big wiggle file if user wants to visually compare RNA-seq data of different library depth.</p></li>
</ul>
</dd>
</dl>
</li>
<li><p><strong>Plots</strong>: Contains all the plots generated by ProkSeq during analysis in pdf, png and tiff format.</p></li>
</ol>
</div></blockquote>
</div>
<div class="section" id="results-output-as-html">
<h2>Results OUTPUT as html<a class="headerlink" href="#results-output-as-html" title="Permalink to this headline">¶</a></h2>
<p>TEST IMAGE</p>
<img alt="_images/volcano.png" src="_images/volcano.png" />
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">ProkSeq</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="INSTALLATION AND DEPENDENCIES.html">INSTALLATION AND DEPENDENCIES:</a></li>
<li class="toctree-l1"><a class="reference internal" href="REQUIREMENTS.html">REQUIRMENTS:</a></li>
<li class="toctree-l1"><a class="reference internal" href="SAMPLES FILE.html">SAMPLE FILE:</a></li>
<li class="toctree-l1"><a class="reference internal" href="SAMPLES FILE.html#data-files">DATA FILES:</a></li>
<li class="toctree-l1"><a class="reference internal" href="PARAMETER FILE.html">PARAMETER FILE:</a></li>
<li class="toctree-l1"><a class="reference internal" href="RUNNING PROKSEQ.html">RUNNING PROKSEQ:</a></li>
<li class="toctree-l1"><a class="reference internal" href="DOCKER.html">DOCKER:</a></li>
<li class="toctree-l1"><a class="reference internal" href="CONDA.html">CONDA:</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">OUTPUT:</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#prokseq-output">ProkSeq OUTPUT</a></li>
<li class="toctree-l2"><a class="reference internal" href="#results-output-as-html">Results OUTPUT as html</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="CONDA.html" title="previous chapter">CONDA:</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, S.nandi, Firoj mahmud.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.4.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/OUTPUT.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>