
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="english">
  <head>
    <meta charset="utf-8" />
    <title>PARAMETER FILE: &#8212; ProkSeq 20.08.2020 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="RUNNING PROKSEQ:" href="RUNNING PROKSEQ.html" />
    <link rel="prev" title="SAMPLE FILE:" href="SAMPLES FILE.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="parameter-file">
<h1>PARAMETER FILE:<a class="headerlink" href="#parameter-file" title="Permalink to this headline">¶</a></h1>
<p><strong>There should be one parameter file. The entries of the file should be as follows.</strong></p>
<p><strong>File &quot;param.yaml&quot;</strong> - definition file in YAML format.</p>
<p>Define the paths and parameters to run the external packages. All the default parameters are considered. However, if users wish to modify any default parameters, please specify here.</p>
<blockquote>
<div><p>NOTE: Any flag/parameter which do not require any value, specify <strong>&quot;TRUE&quot;/&quot;FALSE.&quot;</strong> Quotes are necessary. &quot;<strong>TRUE</strong>&quot; to invoke the flag and &quot;<strong>FALSE</strong>&quot; to suppress.</p>
</div></blockquote>
<p><strong>Describe the Bowtie options below as:</strong></p>
<dl>
<dt>BOWTIE:</dt><dd><p><code class="docutils literal notranslate"><span class="pre">-I</span> <span class="pre">:</span> <span class="pre">0,</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">-X</span> <span class="pre">:</span> <span class="pre">500,</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">-k</span> <span class="pre">:</span> <span class="pre">1,</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">-p</span> <span class="pre">:</span> <span class="pre">1,</span></code></p>
</dd>
</dl>
<p><strong>In case the package salmon is to be run, uncomment(remove #) the following options. These are the default values. These options will overwrite the BOWTIE parameters.</strong></p>
<p><strong>SALMON:</strong></p>
<blockquote>
<div><p><code class="docutils literal notranslate"><span class="pre">SALMONINDEX:-k</span> <span class="pre">:</span> <span class="pre">29</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">SALMONQUANT:</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">-l</span> <span class="pre">:</span> <span class="pre">A-p</span> <span class="pre">:</span> <span class="pre">2</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">--validateMappings</span> <span class="pre">:</span> <span class="pre">&quot;TRUE&quot;</span></code></p>
</div></blockquote>
<p><strong>Define the parameters for AfterQC. Example default parameters are shown as bellow</strong>. All the default parameters are taken. Specify the parameter and value in case of any changes. One cana add the other parameters with value.</p>
<dl>
<dt><strong>AFTERQC:</strong></dt><dd><p><code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">:</span> <span class="pre">35</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">-n</span> <span class="pre">:</span> <span class="pre">5</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">-p</span> <span class="pre">:</span> <span class="pre">35</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">--debubble</span> <span class="pre">:</span> <span class="pre">&quot;TRUE&quot;</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">-a</span> <span class="pre">:</span> <span class="pre">5</span></code></p>
</dd>
</dl>
<p><strong>Define the Featurecounts options as bellow:</strong></p>
<p><strong>FEATURECOUNTS:</strong></p>
<blockquote>
<div><p><code class="docutils literal notranslate"><span class="pre">a</span> <span class="pre">:</span> <span class="pre">oldAnnotationGFF.gtf</span></code> #Define the Featurecounts input GTF file.</p>
<p><code class="docutils literal notranslate"><span class="pre">o</span> <span class="pre">:</span> <span class="pre">FeatCount</span></code> #Define the Featurecounts output file name.</p>
</div></blockquote>
<p><strong>Specify a count file name</strong></p>
<p>COUNTFILE : <strong>countFile.csv</strong></p>
<p><strong>geneBody coverage.r require a bed file. Specify the name of bed file as bellow:</strong></p>
<dl class="simple">
<dt>geneBody_coverage:</dt><dd><p><code class="docutils literal notranslate"><span class="pre">r</span> <span class="pre">:</span> <span class="pre">oldAnnotationGFF.bed</span></code></p>
</dd>
</dl>
<p><strong>Specify if batch effect removal is required. FALSE if not required, else TRUE.</strong></p>
<p>BATCH_EFFECT_REMOVE : &quot;<strong>FALSE</strong>&quot;</p>
<div class="section" id="remove-unwanted-variation-from-rna-seq-data">
<h2>Remove Unwanted Variation from RNA-Seq Data<a class="headerlink" href="#remove-unwanted-variation-from-rna-seq-data" title="Permalink to this headline">¶</a></h2>
<p><strong>To run RUVSeq, uncomment the following two lines.</strong></p>
<dl class="simple">
<dt>RUVSeq:</dt><dd><p><code class="docutils literal notranslate"><span class="pre">nc</span> <span class="pre">:</span> <span class="pre">100</span></code> #Integer - Top 100 genes as ranked by edgeR p-values. Negative control genes to estimate the factors of unwanted variation.</p>
</dd>
</dl>
</div>
<div class="section" id="parameters-for-pathway-analysis">
<h2>Parameters for pathway analysis.<a class="headerlink" href="#parameters-for-pathway-analysis" title="Permalink to this headline">¶</a></h2>
<p><strong>For pathway analysis, define the organism in three alphabets as bellow.</strong></p>
<p>ypy = <em>Yersinia pseudotuberculosis</em></p>
<p>-<strong>User need to change the keg abbreviation of their genome which can be found in</strong> <a class="reference external" href="https://www.genome.jp/kegg/catalog/org_list.html">https://www.genome.jp/kegg/catalog/org_list.html</a>. Here ypy is the <em>Yersinia pseudotuberculosis</em> YPIII</p>
<p><strong>If PATHWAY analysis is not required comment out (add # infront) the PATHWAY section.</strong></p>
<dl class="simple">
<dt>PATHWAY:</dt><dd><p><code class="docutils literal notranslate"><span class="pre">cutoffPositive</span> <span class="pre">:</span> <span class="pre">2.0</span></code>  #logfold upper limit
<code class="docutils literal notranslate"><span class="pre">cutoffNegative</span> <span class="pre">:</span> <span class="pre">-2.0</span></code> #logfold lower limit
<code class="docutils literal notranslate"><span class="pre">Organism</span> <span class="pre">:</span> <span class="pre">ypy</span></code></p>
</dd>
</dl>
<p><strong>For Gene Ontology of the pathway analysis, define GO term and gene name file.</strong></p>
<p>TERM2GENE : data/TERM2GENE.csv</p>
<p>TERM2NAME : data/TERM2NAME.csv</p>
</div>
<div class="section" id="select-the-path-of-tools">
<h2>Select the path of tools<a class="headerlink" href="#select-the-path-of-tools" title="Permalink to this headline">¶</a></h2>
<p><strong>Specify the root path. That means where the ProkSeq bundle is unpacked.</strong></p>
<p>The PATH ROOT : should indicate the current working directory.</p>
<p><strong>The location should have the following folders</strong></p>
<blockquote>
<div><ol class="arabic simple">
<li><p>depend - contains all the binaries of the external packages</p></li>
<li><p>scripts - contains all the modules required for ProkSeq, and prokseq-vx.x.sh</p></li>
<li><p>data - Contains Gene ontology files for pathway analysis.</p></li>
<li><p>Fastq files, and other genome/transcript files.</p></li>
</ol>
</div></blockquote>
<p><strong>If the package is stored in the folder/path /home/user/PROKSEK</strong></p>
<p>PATH ROOT : /home/user/PROKSEK</p>
<p><strong>If the above environment (depend, scripts, data) is true, the following line may be uncommented.</strong></p>
<p>PATH DEFAULT : &quot;TRUE&quot;</p>
<p>PATH SAMTOOLS : /home/user/PROKSEK/depend/samtools/bin #Specify the path to samtools</p>
<p>PATH geneBody_coverage : /home/user/PROKSEK/depend/RSeQC-2.6.2/scripts/  # Specify the path to geneBody_coverage</p>
<p>PATH FEATURECOUNTS : /home/user/PROKSEK/depend/subread-1.4.6-p5-Linux-i386/bin/  #Specify the path to FEATURECOUNTS</p>
<p>PATH FASTQC : /home/user/PROKSEK/depend/FastQC # Specify the path to fastqc</p>
<p>PATH BOWTIE : /home/user/PROKSEK/depend/bowtie2/bowtie2-2.3.5.1-linux-x86_64 #Specify the path to bowtie</p>
<p>PATH SALMON : /home/user/PROKSEK/depend/salmon-latest_linux_x86_64/bin  #Specify the path to salmon if salmon is required</p>
<p>PATH PYPY : /home/user/PROKSEK/depend/pypy2.7-v7.2.0-linux64/bin #Specify the path to pypy required for running afterqc</p>
<p>PATH READFASTA : /home/user/PROKSEK/depend  #Specify the path to readfasta</p>
<p><strong>End of file &quot;param.input&quot;</strong></p>
<p>In general, the entries starting with BOWTIE instructs the program to run the tool with the additional parameter. Similarly SALMON, AFTERQC, FEATURECOUNTS, etc. Entries beginning with PATH indicates the path to the executables of the external tools.  If one uses the bundled packages in the depend folder, <cite>PATH DEFAULT : TRUE</cite> line should be uncommented.</p>
<p><strong>Note: PATH ROOT : path_to_the_current_working_directory should be mentioned.</strong></p>
<p><strong>NOTE: The program will override the Bowtie options, and the package salmon will be run if both bowtie and salmon's options are provided. Therefore, if salmon is required, please comment on the Bowtie option lines, and vice versa.</strong></p>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">ProkSeq</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="INSTALLATION AND DEPENDENCIES.html">INSTALLATION AND DEPENDENCIES:</a></li>
<li class="toctree-l1"><a class="reference internal" href="REQUIREMENTS.html">REQUIRMENTS:</a></li>
<li class="toctree-l1"><a class="reference internal" href="SAMPLES FILE.html">SAMPLE FILE:</a></li>
<li class="toctree-l1"><a class="reference internal" href="SAMPLES FILE.html#data-files">DATA FILES:</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">PARAMETER FILE:</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#remove-unwanted-variation-from-rna-seq-data">Remove Unwanted Variation from RNA-Seq Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#parameters-for-pathway-analysis">Parameters for pathway analysis.</a></li>
<li class="toctree-l2"><a class="reference internal" href="#select-the-path-of-tools">Select the path of tools</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="RUNNING PROKSEQ.html">RUNNING PROKSEQ:</a></li>
<li class="toctree-l1"><a class="reference internal" href="DOCKER.html">DOCKER:</a></li>
<li class="toctree-l1"><a class="reference internal" href="CONDA.html">DOCKER:</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="SAMPLES FILE.html" title="previous chapter">SAMPLE FILE:</a></li>
      <li>Next: <a href="RUNNING PROKSEQ.html" title="next chapter">RUNNING PROKSEQ:</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, S.nandi, Firoj mahmud.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.4.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/PARAMETER FILE.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>