
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="english">
  <head>
    <meta charset="utf-8" />
    <title>SAMPLE FILE: &#8212; ProkSeq 20.08.2020 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="PARAMETER FILE:" href="PARAMETER FILE.html" />
    <link rel="prev" title="REQUIRMENTS:" href="REQUIREMENTS.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="sample-file">
<h1>SAMPLE FILE:<a class="headerlink" href="#sample-file" title="Permalink to this headline">¶</a></h1>
<p>Sample file is required for the program. This file should have the following format. Please don't change the format of the file. Simply replace the fastq, sam and the conditions of the sample.</p>
<p><strong>File &quot;sample&quot; - sample description file.Specify the names of the sample files and tag them as &quot;treat&quot; and &quot;control&quot;.</strong></p>
<p># Specify the genome file, and specify the path where the indexed file will be stored, and the prifex of the indexed genome. Default is 'bowtie2_genome'.</p>
<p>GENOME SequenceChromosome.fasta bowtie2_genome/sequenceChr</p>
<p>-Specify the fastq files</p>
<p>-Specify the output name of the sam files.</p>
<p>-Followed by the tag/class/condition of the sample (treated or control)</p>
<p><strong>List all the fastq files as below.</strong></p>
<p>FASTQ sampleTreat_1.R1.fq sampleTreat_1.R2.fq sampleTreat_1.sam treat</p>
<p>FASTQ sampleTreat_2.R1.fq sampleTreat_2.R2.fq sampleTreat_2.sam treat</p>
<p>FASTQ sampleTreat_3.R1.fq sampleTreat_3.R2.fq sampleTreat_3.sam treat</p>
<p>FASTQ sampleCtrl_1.R1.fq sampleCtrl_1.R2.fq sampleCtrl_1.sam control</p>
<p>FASTQ sampleCtrl_2.R1.fq sampleCtrl_2.R2.fq sampleCtrl_2.sam control</p>
<p>FASTQ sampleCtrl_3.R1.fq sampleCtrl_3.R2.fq sampleCtrl_3.sam control</p>
<p><strong>End of file &quot;sample&quot;</strong></p>
<p>In general, this file starts with GENOME entry. As in the example, the genome file to be used in the analysis is <strong>SequenceChromosome.fasta</strong>. The next argument indicates what would be the prefix of the <strong>indexed genome</strong>, and where to store.&lt;br/&gt;</p>
<p>The following lines are FASTQ. The first argument in the entry is <strong>forward fastq file</strong>, and second is the <strong>reverse fastq</strong> file. <strong>However, if one has only single-end reads, only one entry may be there</strong>.</p>
<p>The subsequent argument is the <strong>SAM</strong> file name. This argument specifies the <strong>output</strong> name of the sam files after running the bowtie. In this example,
<strong>sampleTreat_1.sam</strong>, <strong>sampleTreat_2.sam</strong>, etc. are the sam files for <strong>sampleTreat_1.R1/R2.fq</strong>, <strong>sampleTreat_2.R1/R2.fq</strong>, etc.</p>
<p>The last argument is the condition/class of the sample file (example: &quot;<strong>treat</strong>&quot; and &quot;<strong>control</strong>&quot;).</p>
<div class="section" id="in-case-of-salmon">
<h2>In case of SALMON:<a class="headerlink" href="#in-case-of-salmon" title="Permalink to this headline">¶</a></h2>
<p><strong>File &quot;sample&quot; - sample description file</strong></p>
<p>-Specify the names of the sample files and tag them as &quot;<strong>treat</strong>&quot; and &quot;<strong>control</strong>&quot;.</p>
<p>-Specify the genome file and specify the path where the indexed file will be stored, and the prifex of the indexed genome. Default is 'transcripts_index'.</p>
<p>-<strong>GENOME orf_coding_all.fasta transcripts_index</strong></p>
<p>-Specify the fastq files</p>
<p>-Specify the output name of the quant files.Followed by the tag/class/condition of the sample (<strong>treated</strong> or <strong>control</strong>)</p>
<dl>
<dt>-<strong>List all the fastq files as bellow.</strong></dt><dd><p>FASTQ sampleTreat_1.R1.fq sampleTreat_1.R2.fq sal_quant1 treat</p>
<p>FASTQ sampleTreat_2.R1.fq sampleTreat_2.R2.fq sal_quant2 treat</p>
<p>FASTQ sampleTreat_3.R1.fq sampleTreat_3.R2.fq sal_quant3 treat</p>
<p>FASTQ sampleCtrl_1.R1.fq sampleCtrl_1.R2.fq sal_quant4 control</p>
<p>FASTQ sampleCtrl_2.R1.fq sampleCtrl_2.R2.fq sal_quant5 control</p>
<p>FASTQ sampleCtrl_3.R1.fq sampleCtrl_3.R2.fq sal_quant6 control</p>
</dd>
</dl>
<p><strong>End of file &quot;sample&quot;</strong></p>
<p>In general, this file starts with GENOME entry. As in the example, the transcript file to be used in the analysis is <strong>transcripts.fasta</strong>. The next argument indicates what would be the prefix of the <strong>indexed</strong> file, and where to store.&lt;br/&gt;'</p>
<p>The following lines are FASTQ. The first argument in the entry is <strong>forward fastq</strong> file and second is the <strong>reverse fastq</strong> file. However, if one has only single-end reads, only one entry may be there.</p>
<p>The subsequent argument is the directory where the <strong>alignment file has to be stored</strong>. The last argument is the condition/class of the sample file (example: &quot;<strong>treat</strong>&quot; and &quot;<strong>control</strong>&quot;).</p>
</div>
</div>
<div class="section" id="data-files">
<h1>DATA FILES:<a class="headerlink" href="#data-files" title="Permalink to this headline">¶</a></h1>
<ol class="arabic">
<li><p>Samples files in fastq format.</p></li>
<li><p>Pathway analysis (Optional):</p>
<blockquote>
<div><ol class="arabic">
<li><p>TERN2NAME.csv :Gene Ontology to terms mapping csv file (Eg: <a class="reference external" href="GO:0000001,mitochondrion">GO:0000001,mitochondrion</a> inheritance).This is the GO term classification which is common for all organisms.</p></li>
<li><dl class="simple">
<dt>TERM2GENE.csv :Gene Ontology to gene mapping csv file (Eg: <a class="reference external" href="GO:0003688,YPK_0001">GO:0003688,YPK_0001</a>). This is genome specific Gene ontology file. <strong>TERM2GENE.csv</strong> is a comma delimited 2 column file. First column is the <strong>GO term</strong> and second column is the <strong>gene name</strong>.</dt><dd><dl class="simple">
<dt>:User can download the GO file or GFF annotation file from Genome2D webserver</dt><dd><p>(<a class="reference external" href="http://genome2d.molgenrug.nl/g2d_core_select_genbank.php">http://genome2d.molgenrug.nl/g2d_core_select_genbank.php</a>)</p>
</dd>
</dl>
</dd>
</dl>
</li>
<li><p>For Bowtie implementation:<strong>*Genome file in fasta format</strong></p>
<p>For Salmon implementation: <strong>Transcript file in fasta format</strong></p>
</li>
<li><p>GTF file: Genome annotation in .gtf format. GTF file for bacteria can be downloaded from this link <a class="reference external" href="https://bacteria.ensembl.org/info/data/ftp/index.html">https://bacteria.ensembl.org/info/data/ftp/index.html</a>. Or GFF3 annotation file can be converted to .gtf file by the gff2gtf.sh script provided in ProkSeq folder</p></li>
<li><p>BED file: BED file can be created by using gtf2bed.sh script which is also included in ProkSeq folder</p></li>
</ol>
</div></blockquote>
</li>
</ol>
<p><strong>All these files should be declared in SAMPLE FILE and PARAMETER file.</strong></p>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">ProkSeq</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="INSTALLATION AND DEPENDENCIES.html">INSTALLATION AND DEPENDENCIES:</a></li>
<li class="toctree-l1"><a class="reference internal" href="REQUIREMENTS.html">REQUIRMENTS:</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">SAMPLE FILE:</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#in-case-of-salmon">In case of SALMON:</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#data-files">DATA FILES:</a></li>
<li class="toctree-l1"><a class="reference internal" href="PARAMETER FILE.html">PARAMETER FILE:</a></li>
<li class="toctree-l1"><a class="reference internal" href="RUNNING PROKSEQ.html">RUNNING PROKSEQ:</a></li>
<li class="toctree-l1"><a class="reference internal" href="DOCKER.html">DOCKER:</a></li>
<li class="toctree-l1"><a class="reference internal" href="CONDA.html">DOCKER:</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="REQUIREMENTS.html" title="previous chapter">REQUIRMENTS:</a></li>
      <li>Next: <a href="PARAMETER FILE.html" title="next chapter">PARAMETER FILE:</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, S.nandi, Firoj mahmud.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.4.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/SAMPLES FILE.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>